### PR TITLE
APT-3321 BUGFIX: Unreliable Matching

### PR DESCRIPTION
**Pull request by iurii-provizio**
At least one related bug is fixed: due to leaking of C++ SWIG-wrapped objects in Python wrapper of provizio_dds, publishers could be stockpiling, and when exceeding some numbers (different in different tries, may depend on network bandwidth), newly created publishers fail to match or fail to publish to DDS subscribers. It's due to a limitation of Python when dealing with native (C/C++) code: when Python classes extend C++ classes (wrapped as Python classes using SWIG), then they can't be part of looped references, otherwise Python garbage collector can never clean them up, even when calling `del <object_name>`  and `gc.collect()` directly. It's not that easily reproduced when both publisher and subscriber are on the same machine (hence not detected in unit tests before) likely because with shared memory it can handle much higher numbers publishers & subscribers coexisting.

To test it, you can do the following:
1. Copy (and extract) the attached subscriber test script:
[test_subscriber.zip](https://github.com/provizio/provizio_dds/files/13121881/test_subscriber.zip)
to one machine, and the attached publisher test script:
[test_publisher.zip](https://github.com/provizio/provizio_dds/files/13121883/test_publisher.zip)
to another on the same network.

2. Install "master" version of Python provizio_dds in both machines (first removing the previously installed version, if any):
`python3 -m pip uninstall provizio_dds`
`python3 -m pip install -v git+https://github.com/provizio/provizio_dds.git`

3. Run (quickly enoughm, f.e. over SSH) both the publisher and subscriber, as
`python3 test_publisher.py`
`python3 test_subscriber.py`

4. After some tries it first starts to fail to deliver messages to the subscriber, and then may fail to match (and even crash, on exhausting memory)

5. Now install the fixed version of provizio_dds (first removing the previously installed version):
`python3 -m pip uninstall provizio_dds`
`python3 -m pip install -v git+https://github.com/provizio/provizio_dds.git@feature/APT-3321-Bugfix-Unreliable-Matching`

6. Run the same test as in 3).

7. Wait long enough, the progress will be seen in the output on both side as test numbers (1..10000). If all is fine, it should successfully match and deliver test messages in all 10000 tries on both sides:

```
test_publisher: Successfully matched #10000!
test_publisher: Published #0!
test_publisher: on_has_subscriber False
test_publisher: Successfully unmatched #10000!
```

Please also note, that 2 new unit tests have been added to make sure it doesn't happen - they specifically catch memory leaks in Python both publisher and subscriber. I tried using "master" version of provizio_dds with them and then fixed one, and it works as expected: older one fails with detected memory leaks in the Python publisher, while the new one passes fine.

Additionally, by default Fast DDS does 5 connection tries 0.1 seconds apart each. Then every nodes just passively listens to new connections to be advertised. So in case of just 0.5 seconds connection issue (or busy router, like, with a Pandar on the network) and both participants starting at close time, they would fail to discover each other. I made it 150 tries instead (same 0.1 seconds apart), so it can tolerate up to 15 seconds connectivity issues on start, and also provides much higher chance of at least one advertisement UDP packet to be delivered even in busy networks. I did appropriate tests and see huge improvement in reliability of participants matching.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -322,8 +322,15 @@ if(PYTHON_BINDINGS)
 
     # Make sure Fast-DDS is built prior to SWIG code generation
     if(FAST_DDS_DEPENDENCIES)
-        add_dependencies(fastdds_python_swig_compilation ${FAST_DDS_DEPENDENCIES})
-        add_dependencies(provizio_dds_python_types_swig_compilation ${FAST_DDS_DEPENDENCIES})
+        if(TARGET fastdds_python_swig_compilation)
+            # Older versions of CMake
+            add_dependencies(fastdds_python_swig_compilation ${FAST_DDS_DEPENDENCIES})
+            add_dependencies(provizio_dds_python_types_swig_compilation ${FAST_DDS_DEPENDENCIES})
+        else(TARGET fastdds_python_swig_compilation)
```
Comment (psilar) [inline CMakeLists.txt:329]: Why is `else` the same conditional as `if` ?

**Review by psilar — COMMENTED** [review]
Code compiles clean on my desktop and stack here.  All the `make test`  test failed on both machines, though.  I don't have ros2 installed so that explains some of them but even for the simpllest_pub_sub test I get a `free(): invalid pointer` error.  Probably doesn't matter much as the CI / CD runs these tests anyway.

File: CMakeLists.txt (hunk 1 above)
Comment (iurii-provizio) [reply CMakeLists.txt:329]: @psilar 
Ah, no :) It's a really stupid CMake syntax - every `if` `else` and `endif` can contain the same exact expression, which protects from unintentionally breaking open/close sequence with further changes. It's quite stupid, but standard CMake formatting tools in VS Code add those automatically. I.e. that `else` with expressions is just "else of that specific if". There is `elseif` keyword too in CMake, which may have a condition, but it's that here.

**Comment by iurii-provizio** [comment]
> Code compiles clean on my desktop and stack here. All the `make test` test failed on both machines, though. I don't have ros2 installed so that explains some of them but even for the simpllest_pub_sub test I get a `free(): invalid pointer` error. Probably doesn't matter much as the CI / CD runs these tests anyway.

That's very interesting, it shouldn't be happenning actually. Would be great to know how to reproduce it, as it never happens to me. I'd like to investigate it if I can somehow reproduce the issue.

Also, ROS tests can be disabled when configuring: `cmake -DENABLE_ROS_TESTS=OFF ..`

**Review by oisin-provizio — APPROVED** [review]

**Comment by iurii-provizio** [comment]
Thank you @psilar , the previous run failed seems due to connectivity issues to the build agent jetson (f.e. there is just no logs of the failed run available whatsoever). I've restarted unfinished jobs, which seems to go well so far. If finished clean, I'll merge and update apt_dsp, apt_gui and P&D.

**Pull request merged**